File: alica_engine/src/engine/modelmanagement/factories/TransitionFactory.cpp
#include "engine/modelmanagement/factories/TransitionFactory.h"
#include "engine/model/Transition.h"
#include "engine/modelmanagement/Strings.h"
#include "engine/modelmanagement/factories/KeyMappingFactory.h"
#include "engine/modelmanagement/factories/PreConditionFactory.h"
#include "engine/modelmanagement/factories/VariableBindingFactory.h"

namespace alica
{
Transition* TransitionFactory::create(const YAML::Node& transitionNode, Plan* plan)
{
    std::unique_ptr<KeyMapping> keyMapping;
    if (Factory::isValid(transitionNode[alica::Strings::keyMapping])) {
        keyMapping = KeyMappingFactory::create(transitionNode[alica::Strings::keyMapping]);
    }

    auto* transition = new Transition(std::move(keyMapping));
    Factory::setAttributes(transitionNode, transition);
    Factory::storeElement(transition, alica::Strings::transition);

    if (Factory::isValid(transitionNode[alica::Strings::inState])) {
        Factory::transitionInStateReferences.push_back(
                std::pair<int64_t, int64_t>(transition->getId(), Factory::getReferencedId(transitionNode[alica::Strings::inState])));
    }
    if (Factory::isValid(transitionNode[alica::Strings::outState])) {
        Factory::transitionOutStateReferences.push_back(
                std::pair<int64_t, int64_t>(transition->getId(), Factory::getReferencedId(transitionNode[alica::Strings::outState])));
    }
    if (Factory::isValid(transitionNode[alica::Strings::synchronisation])) {
        Factory::transitionSynchReferences.push_back(
                std::pair<int64_t, int64_t>(transition->getId(), Factory::getReferencedId(transitionNode[alica::Strings::synchronisation])));
    }
    if (Factory::isValid(transitionNode[alica::Strings::condition])) {
        Factory::transitionConditionReferences.push_back(
                std::pair<int64_t, int64_t>(transition->getId(), Factory::getReferencedId(transitionNode[alica::Strings::condition])));
    }

    return transition;
}

void TransitionFactory::attachReferences()
{
    ConditionFactory::attachReferences();
    // transitionOutStateReferences
    for (std::pair<int64_t, int64_t> pairs : Factory::transitionOutStateReferences) {
        Transition* t = (Transition*) Factory::getElement(pairs.first);
        State* st = (State*) Factory::getElement(pairs.second);
        t->setOutState(st);
    }
    Factory::transitionOutStateReferences.clear();
    // transitionInStateReferences
    for (std::pair<int64_t, int64_t> pairs : Factory::transitionInStateReferences) {
        Transition* t = (Transition*) Factory::getElement(pairs.first);
        State* st = (State*) Factory::getElement(pairs.second);
        t->setInState(st);
    }
    Factory::transitionInStateReferences.clear();
    // transitionSynchReferences
    for (std::pair<int64_t, int64_t> pairs : Factory::transitionSynchReferences) {
        Transition* t = (Transition*) Factory::getElement(pairs.first);
        Synchronisation* sync = (Synchronisation*) Factory::getElement(pairs.second);
        t->setSynchronisation(sync);
    }
    Factory::transitionSynchReferences.clear();
    // transitionConditionReferences
    for (std::pair<int64_t, int64_t>& pairs : Factory::transitionConditionReferences) {
        Transition* t = (Transition*) Factory::getElement(pairs.first);
        TransitionCondition* transitionCondition = (TransitionCondition*) Factory::getElement(pairs.second);
        t->setTransitionCondition(transitionCondition);
    }
    Factory::transitionConditionReferences.clear();
}
} // namespace alica

File: alica-supplementary/plan-conversion-tool/src/main/java/com/rapyutarobotics/alica/factories/TransitionFactory.java
package com.rapyutarobotics.alica.factories;

import com.rapyutarobotics.alica.ConversionProcess;
import com.rapyutarobotics.alica.Tags;
import de.unikassel.vs.alica.planDesigner.alicamodel.State;
import de.unikassel.vs.alica.planDesigner.alicamodel.Synchronisation;
import de.unikassel.vs.alica.planDesigner.alicamodel.Transition;
import javafx.util.Pair;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

public class TransitionFactory extends Factory {
    public static Transition create(Element transitionNode, ConversionProcess cp) {
        Transition transition = new Transition();
        Factory.setAttributes(transitionNode, transition);
        cp.addElement(transition);
        cp.transitionInStateReferences.put(transition.getId(),  cp.getReferencedId(transitionNode.getElementsByTagName(Tags.INSTATE).item(0).getTextContent()));
        cp.transitionOutStateReferences.put(transition.getId(),  cp.getReferencedId(transitionNode.getElementsByTagName(Tags.OUTSTATE).item(0).getTextContent()));
        NodeList synchronisationsList = transitionNode.getElementsByTagName(Tags.SYNCHRONISATION);
        if (synchronisationsList.getLength() > 0 ) {
            cp.transitionSynchReferences.put(transition.getId(), cp.getReferencedId(synchronisationsList.item(0).getTextContent()));
        }

        NodeList preConditionList = transitionNode.getElementsByTagName(Tags.PRECONDITION);
        if (preConditionList.getLength() > 0) {
            transition.setPreCondition(PreConditionFactory.create((Element) preConditionList.item(0), cp));
        }

        return transition;
    }

    public static void attachReferences(ConversionProcess cp) {
        ConditionFactory.attachReferences(cp);

        for (Pair<Long, Long> entry : cp.transitionInStateReferences.getEntries()) {
            Transition transition = (Transition) cp.getElement(entry.getKey());
            State state = (State) cp.getElement(entry.getValue());
            if (state == null) {
                throw new RuntimeException("[TransitionFactory] Incoming state with ID " + entry.getValue() + " unknown!");
            }
            transition.setInState(state);
        }
        cp.transitionInStateReferences.clear();

        for (Pair<Long, Long> entry : cp.transitionOutStateReferences.getEntries()) {
            Transition transition = (Transition) cp.getElement(entry.getKey());
            State state = (State) cp.getElement(entry.getValue());
            if (state == null) {
                throw new RuntimeException("[TransitionFactory] Outgoing state with ID " + entry.getValue() + " unknown!");
            }
            transition.setOutState(state);
        }
        cp.transitionOutStateReferences.clear();

        for (Pair<Long, Long> entry : cp.transitionSynchReferences.getEntries()) {
            Transition transition = (Transition) cp.getElement(entry.getKey());
            Synchronisation synchronisation = (Synchronisation) cp.getElement(entry.getValue());
            if (synchronisation == null) {
                throw new RuntimeException("[TransitionFactory] Synchronisation with ID " + entry.getValue() + " unknown!");
            }
            transition.setSynchronisation(synchronisation);
        }
        cp.transitionSynchReferences.clear();
    }
}
